### PR TITLE
1265: Workaround empty secure storage

### PR DESCRIPTION
**Pull request by f1sh1918**
### Short description

We are facing some not reproducible issue on android devices that the user data (stored cards) can't be read from secure storage after app restart. This relies on an issue from `flutter secure storage`
https://github.com/mogol/flutter_secure_storage/issues/653
Some devs report that a fake read could solve this issue.

### Proposed changes
- add a fake read for android devices
- add a second read for android devices that may contain the user data
- report error to sentry if first read is null but second read contains data to collect more information

### Side effects

- the initialization of the cards may be a bit slower for android but it shouldn't be a big performance issue

### Resolved issues

related to #1265

**Note:** We have to contact users that facing the issue if it was solved by this workaround, since we can not really test it
Just test if a saved card will be restored/loaded (still) properly

### CODE REVIEW

File: frontend/lib/identification/user_code_store.dart [hunk 1]
```diff
@@ -24,7 +28,19 @@ class UserCodeStore {
...
-    final String? userCodesBase64 = await storage.read(key: _userCodesBase64Key);
+    String? userCodesBase64;
+    // make fake read on android - workaround for https://github.com/mogol/flutter_secure_storage/issues/653
+    if (Platform.isAndroid) {
+      String? userCodesBase64FirstRead = await storage.read(key: _userCodesBase64Key);
+      userCodesBase64 = await storage.read(key: _userCodesBase64Key);
+      // report error if first read is null but second read contains data to collect sentry data for reproduction
+      if (userCodesBase64FirstRead == null && userCodesBase64 != null) {
+        reportError('First read from secure storage is null but user data exists', null);
+      }
```
Comment (michael-markl) [inline frontend/lib/identification/user_code_store.dart:39]: Maybe, we can report more potential misbehavior here to get a better understanding.
Maybe something like this:
```suggestion
      if (userCodesBase64FirstRead != userCodesBase64) {
        bool firstIsNull = userCodesBase64FirstRead == null;
        bool secondIsNull = userCodesBase64 == null;
        reportError("First read from secure storage differs from second: firstNull=${firstIsNull}, secondNull=${secondIsNull}", null);
      }
```

File: frontend/lib/identification/user_code_store.dart [hunk 2]
```diff
@@ -24,7 +28,19 @@ class UserCodeStore {
...
+    } else {
+      userCodesBase64 = await storage.read(key: _userCodesBase64Key);
+    }
+
     if (userCodesBase64 == null) return [];
     return userCodesBase64
         .split(_storageDelimiter)
```
Comment (michael-markl) [inline frontend/lib/identification/user_code_store.dart:46]: I have a question regarding the `importLegacyCard` function below. 

It constructs `UserCodeModel` without initializing it and then directly runs `insertCode` which in turn runs `UserCodeStore.store()`.
This seems quite bad.

I think it is not the cause of the problem, as `insertCode` is only called when a legacy code is found and thus should only be called once ever from importLegacyCard. But we should definitely refactor this.
`importLegacyCard` should return `Future<DynamicUserCode?>` and there should be a separate function `deleteLegacyCard` which should delete the key. The `initialize`  function of `UserCodeModel` should take care of calling these functions.

Also, we should add assertions to the functions of UserCodeModel which ensure they can only be called if UserCodeModel is initialized.

**Review by michael-markl — APPROVED** [review]

File: frontend/lib/identification/user_code_store.dart (hunk 2 above)
Comment (f1sh1918) [reply frontend/lib/identification/user_code_store.dart:48]: i'm gonna create a task for it, thx for the hint

**Comment by f1sh1918** [comment]
Can anyone review AND test it :) @sarahsporck and @ztefanie

**Review by sarahsporck — APPROVED** [review]
Tested in Android emulator. Seems to work the same way as before 👍🏼

**Pull request merged**